Build iota.c with MS Visual-Studio and  PR: Rest api restructure #1577

DerAnhalter opened the pull request:
# Description of change

Code changes to be able to compile iota.c to Microsoft Visual-Studio.


## Type of change

Choose a type of change, and delete any options that are not relevant.

- Enhancement (a non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

Running the examples

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] My code follows the contribution guidelines for this project
- [x ] I have performed a self-review of my own code
- [ x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests using CTest that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

DerAnhalter committed:
Microsoft Visual-Studio: Problems with multiple Files with same Name

DerAnhalter committed:
Errors when Building iota.c with Microsoft Visual-Studio (#2)

DerAnhalter committed:
Example - node_info does not return Information about the Node. (#3)

DerAnhalter committed:
Example - Wallet - create_alias_output fails (#4)

oopsmonk commented on src/core/utils/uint256.c:
```diff
@@ -134,20 +134,26 @@ uint256_t *uint256_from_hex_str(char const *str) {
   }
 
   // create temporary string with maximum 256 bit hex length (4 * uint64_t)
-  char str_max_temp[str_max_len];
```
MSVC doesn't support the variable-length array(VLA), this feature is important for embedded projects to avoid problems of dynamic memory allocation. 
we can use the compiler's predefined macro to solve this issue

```c
#if defined(_MSC_VER)
// non-VAL implementation for MS Visual-Studio
#else 
// the original VLA implementation
#endif 
```

DerAnhalter replied:
I have changed the code: now using compiler's predefined macro for solving the issue of variable-length array

oopsmonk commented on src/crypto/iota_crypto.c:
```diff
@@ -142,11 +142,27 @@ int iota_crypto_sign(uint8_t const priv_key[], uint8_t msg[], size_t msg_len, ui
 
 int iota_crypto_sign_open(uint8_t msg[], size_t msg_len, uint8_t const pub_key[], uint8_t signature[]) {
 #if defined(CRYPTO_USE_SODIUM)
-  uint8_t sm_buff[ED_SIGNATURE_BYTES + msg_len];
```
see the above suggestion to keep the VAL implementation.

oopsmonk reviewed (changes requested):
thanks for the PR, please fix the coding style and suggestions.
I think the coding style can be fixed via the [ClangFormat](https://marketplace.visualstudio.com/items?itemName=LLVMExtensions.ClangFormat) on Visual Studio IDE.

DerAnhalter committed:
MSVC doesn't support the variable-length array(VLA). Now using the
compiler's predefined macro.

DerAnhalter committed:
Problems with multiple Files with same Name:
There is no need to rename the Sourcefiles. I can define an individual
Name for the Objectfile.

DerAnhalter committed:
Coding style changes

DerAnhalter commented:
Hello,
i have made these modifications:
- Using compiler's predefined macro for solving the issue of  variable-length array
- fix the coding style. Thank you for the Tip (ClangFormat Plugin).
- There is no need to rename the Sourcefiles. I can define an individual Name for the Objectfile.

DerAnhalter committed:
ninja_clang_sodium-test, fix: runtime error: member access within null
pointer of type 'UT_array'

DerAnhalter committed:
coding style...

DerAnhalter committed:
ninja_clang_sodium-test, fix: runtime error: member access within null
pointer of type 'UT_array'

DerAnhalter committed:
Is the plugin array part of the info response anymore ?

DerAnhalter commented:
Is the plugin array part of the info response anymore ?

Here a Answer from my Node:

`{"name":"HORNET","version":"b3189432","status":{"isHealthy":true,"latestMilestone":{"index":78509,"timestamp":1656269558,"milestoneId":"0x84e08a27239963c6b98fa282127a0234509c6cfbec5edac37681583b42b26443"},"confirmedMilestone":{"index":78509,"timestamp":1656269558,"milestoneId":"0x84e08a27239963c6b98fa282127a0234509c6cfbec5edac37681583b42b26443"},"pruningIndex":42400},"supportedProtocolVersions":[2],"protocol":{"version":2,"networkName":"alphanet-7","bech32HRP":"rms","minPoWScore":1000,"belowMaxDepth":15,"rentStructure":{"vByteCost":500,"vByteFactorData":1,"vByteFactorKey":10},"tokenSupply":"2779530283277761"},"pendingProtocolParameters":[],"baseToken":{"name":"Shimmer","tickerSymbol":"SMR","unit":"SMR","subunit":"glow","decimals":6,"useMetricPrefix":false},"metrics":{"blocksPerSecond":2.2,"referencedBlocksPerSecond":2,"referencedRate":90.9090909090909},"features":["PoW"]}`

oopsmonk commented:
> Is the plugin array part of the info response anymore ?
> 
it was removed at [HORNET-2.0.0-alpha.21](https://github.com/iotaledger/hornet/releases/tag/v2.0.0-alpha.21)

DerAnhalter commented:
>> Is the plugin array part of the info response anymore ?
> 
> it was removed at HORNET-2.0.0-alpha.21
> 

I think that it is not critical when info about plugins is not available. I therefore propose to tolerate the absence of the plugin-array in the deser_node_info function.

DerAnhalter committed:
Tolerating the absence of the plugin-array in the deser_node_info function.

oopsmonk commented:
> 

looks good to me, I will remove the plugin element later.
